intermediate-goals.Rmd

```diff
@@ -0,0 +1,148 @@
+-   How can I define rules to operate on similar files? 
+    -   Pattern rules
+-   How can I eliminate redundancy in my Makefiles?
+    -   Variables
+    -   Functions
+-   How should I document a Makefile?
+-   What other tools can I use?
+    -   drake for R
+    -   snakemake for Python
+
+## Reusable software
+
+Based on <https://merely-useful.github.io/still-magic/en/configure.html>
+and <https://merely-useful.github.io/still-magic/en/logging.html>
+
+-   How do I parse the command line from within a Python script?
+    -   basic structure of python scripts
+    -   argparse
+-   How would I implement a configuration file?
+    -   as a data structure (JSON)
+    -   as key/value pairs (YAML)
+-   How would I report my program's progress?
+    -   logging levels
+    -   logging channels
+    -   configuration
+-   How do I document my programs?
+    -   documentation generators
+    -   docstrings in Python
+    -   roxygen in R
+
+## Version control
+
+-   How do I do multiple things at once?
+    -   branches
+    -   branch naming
+    -   branch management conventions (like feature branches)
+-   How do I suggest changes to other people's code?
+    -   forking repositories
+    -   creating pull requests
+    -   reviewing pull requests
+
+## Testing
+
+Based on <http://katyhuff.github.io/python-testing/>,
+and <https://merely-useful.github.io/still-magic/en/unit.html>
+
+-   Why should I test my code?
+-   How do I implement tests/checks at specific points within my code?
+    -   assertions (brief, this is covered in novice course)
+    -   exceptions
+-   How do I test the overall behaviour of parts of my code?
+    -   unit testing
+-   How can I run commands automatically every time a repository is updated?
+    -   continuous integration
+    -   Travis
+-   How can I tell what I have and haven't tested?
+    -   coverage
+
+## Packaging
+
+Based on <https://carpentrieslab.github.io/python-aos-lesson/01-conda/>
+and <https://merely-useful.github.io/still-magic/en/package.html>
+
+-   How do I install and manage my Python environment?
+    -   Python distributions (Anaconda)
+    -   Package managers (pip, conda)
+    -   conda environments
+-   How can I write a package that can be installed via pip?
+-   How do I write a package that can be shared on CRAN?
+-   How do version numbers work?
+
+## Project management
+
+-   How can **we** keep track of who's doing what and what's still to be done?
+    -   issue tracking
+    -   issue tagging as lightweight workflow
+-   How can **we** figure out what to build next?
+    -   planning vs. agile
+    -   3x3 prioritization
+-   How can **we** make decisions?
+    -   how to run a meeting
+    -   project governance
+    -   strategies for handling difficult project members
+
+## What have we left out?
+
+-   Object-oriented programming
+-   Nonstandard evaluation (in R)
+-   Refactoring
+-   Profiling and performance tuning
```

Comment:
This could be briefly explained conceptually with links to resources for those interested.